go/vt/vtctl/reparentutil/durability_policies.go

```diff
@@ -0,0 +1,238 @@
+// IsReplicaSemiSync returns the replica semi-sync setting from the tablet record.
+// Prefer using this function if tablet record is available.
+func IsReplicaSemiSync(primary, replica *topodatapb.Tablet) bool {
+	curDurabilityPolicyMutex.Lock()
+	defer curDurabilityPolicyMutex.Unlock()
+	return curDurabilityPolicy.isReplicaSemiSync(primary, replica)
+}
+
+//=======================================================================
+
+// durabilityNone has no semi-sync and returns NeutralPromoteRule for Primary and Replica tablet types, MustNotPromoteRule for everything else
+type durabilityNone struct{}
+
+func (d *durabilityNone) promotionRule(tablet *topodatapb.Tablet) promotionrule.CandidatePromotionRule {
+	switch tablet.Type {
+	case topodatapb.TabletType_PRIMARY, topodatapb.TabletType_REPLICA:
+		return promotionrule.Neutral
+	}
+	return promotionrule.MustNot
+}
+
+func (d *durabilityNone) semiSyncAckers(tablet *topodatapb.Tablet) int {
+	return 0
+}
+
+func (d *durabilityNone) isReplicaSemiSync(primary, replica *topodatapb.Tablet) bool {
+	return false
+}
+
+//=======================================================================
+
+// durabilitySemiSync has 1 semi-sync setup. It only allows Primary and Replica type servers to acknowledge semi sync
+// It returns NeutralPromoteRule for Primary and Replica tablet types, MustNotPromoteRule for everything else
+type durabilitySemiSync struct{}
+
+func (d *durabilitySemiSync) promotionRule(tablet *topodatapb.Tablet) promotionrule.CandidatePromotionRule {
+	switch tablet.Type {
+	case topodatapb.TabletType_PRIMARY, topodatapb.TabletType_REPLICA:
+		return promotionrule.Neutral
+	}
+	return promotionrule.MustNot
+}
+
+func (d *durabilitySemiSync) semiSyncAckers(tablet *topodatapb.Tablet) int {
+	return 1
+}
+
+func (d *durabilitySemiSync) isReplicaSemiSync(primary, replica *topodatapb.Tablet) bool {
+	switch replica.Type {
+	case topodatapb.TabletType_PRIMARY, topodatapb.TabletType_REPLICA:
+		return true
+	}
+	return false
+}
+
+//=======================================================================
+
+// durabilityCrossCell has 1 semi-sync setup. It only allows Primary and Replica type servers from a different cell to acknowledge semi sync.
+// This means that a transaction must be in two cells for it to be acknowledged
+// It returns NeutralPromoteRule for Primary and Replica tablet types, MustNotPromoteRule for everything else
+type durabilityCrossCell struct{}
+
+func (d *durabilityCrossCell) promotionRule(tablet *topodatapb.Tablet) promotionrule.CandidatePromotionRule {
+	switch tablet.Type {
+	case topodatapb.TabletType_PRIMARY, topodatapb.TabletType_REPLICA:
+		return promotionrule.Neutral
+	}
+	return promotionrule.MustNot
+}
+
+func (d *durabilityCrossCell) semiSyncAckers(tablet *topodatapb.Tablet) int {
+	return 1
+}
+
+func (d *durabilityCrossCell) isReplicaSemiSync(primary, replica *topodatapb.Tablet) bool {
+	// Prevent panics.
+	if primary.Alias == nil || replica.Alias == nil {
```

Comment:
Don't we need this check in `durabilitySemiSync` as well?

Reply:
I'll move that check to the Exported function which calls these methods, so that they can always assume that the provided values are non-negative